### PR TITLE
add memcached auto-discovery support from thanos

### PR DESCRIPTION
**Pull request by roystchiang**
Signed-off-by: Roy Chiang <roychi@amazon.com>

**What this PR does**:
Updates Thanos to a version that includes memcached auto-discovery support, introduced in https://github.com/thanos-io/thanos/pull/4487.

Auto-discovery is a way for clients to automatically discover nodes belonging to a memcached cluster. It is different from DNS discovery because the node resolution is not done on the DNS level, but rather through talking to memcached.
* https://cloud.google.com/memorystore/docs/memcached/auto-discovery-overview
* https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.HowAutoDiscoveryWorks.html

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by alolita** [comment]
Can you please also review this PR @alolita @aneurysm9

File: pkg/storage/tsdb/memcache_client_config.go [hunk 1]
```diff
@@ -29,6 +30,7 @@ func (cfg *MemcachedClientConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefi
 	f.IntVar(&cfg.MaxGetMultiConcurrency, prefix+"max-get-multi-concurrency", 100, "The maximum number of concurrent connections running get operations. If set to 0, concurrency is unlimited.")
 	f.IntVar(&cfg.MaxGetMultiBatchSize, prefix+"max-get-multi-batch-size", 0, "The maximum number of keys a single underlying get operation should run. If more keys are specified, internally keys are split into multiple batches and fetched concurrently, honoring the max concurrency. If set to 0, the max batch size is unlimited.")
 	f.IntVar(&cfg.MaxItemSize, prefix+"max-item-size", 1024*1024, "The maximum size of an item stored in memcached. Bigger items are not stored. If set to 0, no maximum size is enforced.")
+	f.BoolVar(&cfg.AutoDiscovery, prefix+"auto-discovery", false, "Use memcached auto-discovery instead of DNS resolution")
```
Comment (bboreham) [inline pkg/storage/tsdb/memcache_client_config.go:33]: Can we get a bit more descriptive than "memcached auto-discovery"?
I Googled that phrase and am not sure I found what you meant.

**Review by bboreham — COMMENTED** [review]
Thanks for this PR.
Since you're upgrading thanos, is this a new feature?  If so please link to the upstream details in the PR, maybe also in the CHANGELOG.
Please also check what other changes are brought in by this upgrade (maybe not many, looks like 6 days apart) and state what you found, again in the PR description and CHANGELOG if appropriate.

File: pkg/storage/tsdb/memcache_client_config.go (hunk 1 above)
Comment (jeromeinsf) [reply pkg/storage/tsdb/memcache_client_config.go:33]: https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.HowAutoDiscoveryWorks.html

Comment (alvinlin123) [reply pkg/storage/tsdb/memcache_client_config.go:33]: There is also: https://cloud.google.com/memorystore/docs/memcached/auto-discovery-overview

Comment (roystchiang) [reply pkg/storage/tsdb/memcache_client_config.go:33]: I have updated the PR description to include the thanos PR link. Also added 2 links for explaining what auto-discovery means for memcached.

There are no other changes brought by upgrading Thanos in this PR. 

Please let me know if you have other concerns.

Comment (alvinlin123) [reply pkg/storage/tsdb/memcache_client_config.go:33]: @bboreham  do you have anymore concern on this PR?

**Review by alvinlin123 — COMMENTED** [review]
Since the auto_discovery feature here is Cloud provider specific, I think we should add some document like 

https://cortexmetrics.io/docs/configuration/arguments/#dns-service-discovery 

or maybe in 

https://cortexmetrics.io/docs/chunks-storage/aws-tips/

File: pkg/storage/tsdb/memcache_client_config.go (hunk 1 above)
Comment (bboreham) [reply pkg/storage/tsdb/memcache_client_config.go:33]: The Changelog entry is circular and the Thanos link doesn't explain auto-discovery.
I think you're saying it is an ElasticCache feature that is also supported by Google Cloud; the Cortex docs should link to one or both of those if there is nothing more neutral.

**Review by bboreham — APPROVED** [review]
Thanks!

**Review by alvinlin123 — APPROVED** [review]

**Pull request merged**